validator/src/main/java/org/ethereum/beacon/validator/proposer/BeaconChainProposerImpl.java

```diff
@@ -0,0 +1,177 @@
+    BeaconBlock newBlock = builder.build();
+    BeaconState newState = blockTransition.apply(newBlock, state);
+    builder.withStateRoot(specHelpers.hash_tree_root(newState));
+
+    // sign off on proposal
+    BeaconBlock blockWithoutSignature = builder.build();
+    Bytes96 signature = getProposalSignature(state, blockWithoutSignature, signer);
+    builder.withSignature(signature);
+
+    return builder.build();
+  }
+
+  private Bytes96 getProposalSignature(
+      BeaconState state, BeaconBlock blockWithoutSignature, MessageSigner<Bytes96> signer) {
+    ProposalSignedData proposal =
+        new ProposalSignedData(
+            state.getSlot(),
+            chainSpec.getBeaconChainShardNumber(),
+            specHelpers.hash_tree_root(blockWithoutSignature));
+    Hash32 proposalRoot = specHelpers.hash_tree_root(proposal);
+    Bytes8 domain = specHelpers.get_domain(state.getForkData(), state.getSlot(), PROPOSAL);
+    return signer.sign(proposalRoot, domain);
+  }
+
+  private Bytes96 getRandaoReveal(BeaconState state, UInt24 index, MessageSigner<Bytes96> signer) {
+    ValidatorRecord proposer = state.getValidatorRegistry().get(SpecHelpers.safeInt(index));
+    Hash32 hash = Hash32.wrap(Bytes32.leftPad(proposer.getProposerSlots().toBytesBigEndian()));
+    Bytes8 domain = specHelpers.get_domain(state.getForkData(), state.getSlot(), RANDAO);
+    return signer.sign(hash, domain);
+  }
+
+  /*
+   Let D be the set of Eth1DataVote objects vote in state.eth1_data_votes.
+
+   If D is empty:
+     Let block_hash be the block hash of the ETH1_FOLLOW_DISTANCEth ancestor of the head of the
+       canonical eth1.0 chain.
+     Let deposit_root be the deposit root of the eth1.0 deposit contract at the block defined by
+       block_hash.
+
+   If D is nonempty:
+     Let best_vote be the member of D that has the highest vote.eth1_data.vote_count,
+       breaking ties by favoring block hashes with higher associated block height.
+     Let block_hash = best_vote.eth1_data.block_hash.
+     Let deposit_root = best_vote.eth1_data.deposit_root.
+
+   Set block.eth1_data = Eth1Data(deposit_root=deposit_root, block_hash=block_hash).
+  */
+  private Eth1Data getEth1Data(BeaconState state) {
+    List<Eth1DataVote> eth1DataVotes = state.getEth1DataVotes();
+    Eth1Data contractData = depositContract.getEth1Data(chainSpec.getEth1FollowDistance());
+
+    if (eth1DataVotes.isEmpty()) {
+      return contractData;
+    }
+
+    UInt64 maxVotes =
+        eth1DataVotes.stream().map(Eth1DataVote::getVoteCount).max(UInt64::compareTo).get();
+    Eth1DataVote bestVote =
+        eth1DataVotes.stream()
+            .filter(eth1DataVote -> eth1DataVote.getVoteCount().equals(maxVotes))
+            .reduce((first, second) -> second)
+            .get();
+
+    // verify best vote data and return if verification passed,
+    // otherwise, return data from contract
+    return depositContract
+        .getDepositRoot(bestVote.getEth1Data().getBlockHash())
```

Comment:
why not just `bestVote.getEth1Data().getDepositRoot()` ?

Reply:
I think that proposer should get a proof that root posted by a 3rd party exists.

Reply:
Oh, got it!